sbncode/CAFMaker/FillTrue.cxx

```diff
@@ -127,6 +130,190 @@ namespace caf {
+
+        TVector3 loc_nosce_v(loc_nosce.x(), loc_nosce.y(), loc_nosce.z());
+        
+        // Match to MC-trajectory
+        TVector3 direction;
+        float closest_dist = -1.;
+        int traj_index = -1;
+        for (unsigned i_traj = 0; i_traj < particle.NumberTrajectoryPoints(); i_traj++) {
+          if (closest_dist < 0. || (particle.Position(i_traj).Vect() - loc_nosce_v).Mag() < closest_dist) {
+            direction = particle.Momentum(i_traj).Vect().Unit();
+            closest_dist = (particle.Position(i_traj).Vect() - loc_nosce_v).Mag();
+            traj_index = i_traj;
+          }
+        }
+
+        // residual range
+        truep.rr = 0;
+        if (traj_index >= 0) {
+          for (int i_traj = traj_index+1; i_traj < (int)particle.NumberTrajectoryPoints(); i_traj++) {
+            truep.rr += (particle.Position(i_traj).Vect() - particle.Position(i_traj-1).Vect()).Mag();
+          }
+
+          // Also account for the distance from the Hit point to the matched trajectory point
+          double hit_distance_along_particle = (loc_nosce_v - particle.Position(traj_index).Vect()).Dot(particle.Momentum(traj_index).Vect().Unit());
+          truep.rr += -hit_distance_along_particle;
+        }
+        else truep.rr = -1;
+
+        // pitch
+        //
+        // The true pitch is affected by the presence of space charge, so we cannot
+        // directly apply the true direction. We include the effect of space charge
+        // on the true pitch here
+        if (closest_dist >= 0. && direction.Mag() > 1e-4) {
+          geo::PlaneID plane(srtrack.producer, p.tpc, iplane);
+          float angletovert = geo->WireAngleToVertical(geo->View(plane), plane) - 0.5*::util::pi<>();
+          TVector3 loc_mdx_v = loc_nosce_v - direction * (geo->WirePitch(geo->View(plane)) / 2.);
+          TVector3 loc_pdx_v = loc_nosce_v + direction * (geo->WirePitch(geo->View(plane)) / 2.);
+
+          // Convert types for helper functions
+          geo::Point_t loc_mdx(loc_mdx_v.X(), loc_mdx_v.Y(), loc_mdx_v.Z());
+          geo::Point_t loc_pdx(loc_pdx_v.X(), loc_pdx_v.Y(), loc_pdx_v.Z());
+
+          // Map to wires
+          if (sce && sce->EnableSimSpatialSCE()) { 
+            int corr = geo->TPC(plane.TPC).DriftDir()[0];
+
+            geo::Vector_t offset_m = sce->GetPosOffsets(loc_mdx);
+            offset_m.SetX(offset_m.X()*corr); // convert from drift direction to detector direction
+            loc_mdx = loc_mdx + offset_m;
+
+            geo::Vector_t offset_p = sce->GetPosOffsets(loc_pdx);
+            offset_p.SetX(offset_p.X()*corr); // convert from drift direction to detector direction
+            loc_pdx = loc_pdx + offset_p;
+          }
+
+          // Direction at wires
+          geo::Vector_t dir = (loc_pdx - loc_mdx) /  (loc_mdx - loc_pdx).r();
+
+          // Pitch at wires
+          double cosgamma = std::abs(std::sin(angletovert)*dir.Y() + std::cos(angletovert)*dir.Z());
```

Comment:
```suggestion
          double cosgamma = std::abs(plane.SinPhiZ()*dir.Y() + plane.CosPhiZ()*dir.Z());
```
Short explanation: `geo::WireAngleToVertical()` [is actually not the wire angle to vertical](https://cdcvs.fnal.gov/redmine/issues/27790), but it is the wire angle (equivalent to `geo::PlaneGeo::ThetaZ()`), and that is the reason while in the definition a few lines above in your code its value was rotated by a right angle. `geo::PlaneGeo` also provides the direction orthogonal to the wires (i.e. the direction that the wire plane measures), calling it `PhiZ()`, and since _z_ is "the horizontal" and _y_ is "the vertical", `PhiZ()` happens to be also the angle _of the wire_ from the vertical (barred possibly a sign, which in `PhiZ()` is defined as the direction where the coordinate increases).
Finally, the plane provides pre-computed sine and cosine.

Reply:
See above comment. I agree that the existing code is kind of confusing. But it gives the correct result and is the same interface that the Calorimetry uses.